src/poll.rs

```diff
@@ -0,0 +1,258 @@
+                if let Some(token) = self.tokens.get(&(event.userdata as i32)) {
+                    if let Some(status) = status_map.get_mut(token) {
+                        *status |= event.type_.raw();
+                    } else {
+                        status_map.insert(*token, event.type_.raw());
+                    }
+                }
+            }
+        }
+        for (token, status) in status_map.into_iter() {
+            events.push(Event {
+                token,
+                status: Interest::from_bits(status).unwrap(),
+            });
+        }
+        Ok(events)
+    }
+}
+
+/// Structs for containting poll fds
+#[derive(Clone, Debug)]
+pub struct Selector {
+    poll_fds: HashMap<i32, PollFd>,
+}
+
+impl Selector {
+    pub fn new() -> Selector {
+        Selector {
+            poll_fds: HashMap::new(),
+        }
+    }
+}
+
+impl Default for Selector {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Interest used in registering
+///
+/// Currently only support read, write, and both(read and write).
+#[derive(Copy, Clone, Debug)]
+pub enum Interest {
+    Read,
+    Write,
+    Both,
+}
+
+impl Interest {
+    pub fn from_bits(bits: u8) -> Option<Interest> {
+        match bits {
+            1 => Some(Interest::Read),
+            2 => Some(Interest::Write),
+            3 => Some(Interest::Both),
+            _ => None,
+        }
+    }
+
+    pub fn is_readable(&self) -> bool {
+        matches!(self, Interest::Read | Interest::Both)
+    }
+
+    pub fn is_writable(&self) -> bool {
+        matches!(self, Interest::Write | Interest::Both)
+    }
+}
+
+/// Pollable file descriptor
+#[derive(Copy, Clone, Debug)]
+pub struct PollFd {
+    fd: RawFd,
+    interest: Interest,
+}
+
+impl AsPollFd for PollFd {
+    fn push_sub(&self, subs: &mut Vec<Subscription>) {
+        match self.interest {
+            Interest::Read => {
+                subs.push(Subscription {
+                    userdata: self.fd as u64,
+                    u: SubscriptionU {
+                        tag: EVENTTYPE_FD_READ.raw(),
+                        u: SubscriptionUU {
+                            fd_read: SubscriptionFdReadwrite {
+                                file_descriptor: self.fd as u32,
+                            },
+                        },
+                    },
+                });
+            }
+            Interest::Write => {
+                subs.push(Subscription {
+                    userdata: self.fd as u64,
+                    u: SubscriptionU {
+                        tag: EVENTTYPE_FD_WRITE.raw(),
+                        u: SubscriptionUU {
+                            fd_read: SubscriptionFdReadwrite {
+                                file_descriptor: self.fd as u32,
+                            },
+                        },
+                    },
+                });
+            }
+            Interest::Both => {
+                subs.push(Subscription {
```

Comment:
@hydai [wasmedge_wasi poll_oneoff](https://github.com/WasmEdge/WasmEdge/blob/e532fdfba2ccdc5f8fcd89075a13620476fe9861/lib/host/wasi/inode-linux.cpp#L1390) will return __WASI_ERRNO_EXIST. Because the  FD registered by EVENTTYPE_FD_READ.

Reply:
Is this an expected behavior? Or is it a wrong implementation in WasmEdge WASI-Socket?

Reply:
A  wrong implementation in WasmEdge WASI.

In linux , WasmEdge WASI use `epoll` to implement `poll_oneoff`. 
When FD subscribe `EVENTTYPE_FD_READ ` and `EVENTTYPE_FD_WRITE`, 
[Poller::read](https://github.com/WasmEdge/WasmEdge/blob/e532fdfba2ccdc5f8fcd89075a13620476fe9861/lib/host/wasi/inode-linux.cpp#L1366) will add the `fd` into `epoll_fd` first, this is ok.
And then, [Poller::write](https://github.com/WasmEdge/WasmEdge/blob/e532fdfba2ccdc5f8fcd89075a13620476fe9861/lib/host/wasi/inode-linux.cpp#L1390) add the `fd` into `epoll_fd` again, it will return error.
If the `fd`  has registered, poller::write should use `EPOLL_CTL_MOD`    :(

Reply:
I created an issue related to this. Will be fixed this week.